### PR TITLE
Support group chain ability

### PR DESCRIPTION
**Pull request by ToWorld**
## Description

What is the purpose of the change?
Description, as follows:
One Chain, only allowed nodes can access the network
and broadcast tx and block and etc..
Use contract to maintain the white list.

Fixes # (issue)

## Type of change

- [ ] New feature (non-breaking change which adds functionality)

## Brief of your solution

Use contract to keep whitelist.
The contract as a kernel contract, which is configured in genesis file.
case1:When receive data from peer node:
Check if the peer node is in whitelist
case2:When broadcast data to peer node:
Check if the peer node is in whitelist

## How Has This Been Tested?

Regression Test

### CODE REVIEW

File: core/core/group_chain.go [hunk 1]
```diff
@@ -0,0 +1,117 @@
...
+	StreamCache map[string]string
+	// key: bcname value: map[string]bool
+	StreamContractCache map[string]map[string]bool
+	// key: bcname value: bool
+	ChainContractCache map[string]bool
+	Mutex              *sync.Mutex
+}
+
+// CheckGroupChain 判断某条链下的某个节点是否在白名单中
+func (xm *XChainMG) CheckGroupChain(bcname, remotePeerID string) bool {
```
Comment (yucaowang) [inline core/core/group_chain.go]: I think 'FindNode' is more clear than 'CheckGroupChain'.

Comment (yucaowang) [reply core/core/group_chain.go]: Or 'FindPeer'

File: core/utxo/group_chain.go [hunk 2]
```diff
@@ -0,0 +1,105 @@
...
+		ResourceLimits: contract.MaxLimits,
+		ContractName:   contractName,
+	}
+	vm, err := uv.vmMgr3.GetVM(moduleName)
+	if err != nil {
+		return 400, nil, err
+	}
+	ctx, err := vm.NewContext(contextConfig)
+	if err != nil {
+		return 400, nil, err
```
Comment (fxsjy) [inline core/utxo/group_chain.go]: ctx not release here?

File: core/core/group_chain.go [hunk 3]
```diff
@@ -0,0 +1,117 @@
...
+	xm.groupChainCache.Mutex.Lock()
+	defer xm.groupChainCache.Mutex.Unlock()
+	xm.groupChainCache.StreamCache = data
+}
+
+func (xm *XChainMG) updateGroupChainCache() {
+	for {
+		xm.updateContractCache()
+		xm.updateStreamCache()
+		time.Sleep(2 * time.Second)
```
Comment (fxsjy) [inline core/core/group_chain.go]: magic number, better to be a const or flag

File: core/p2p/p2pv2/server.go [hunk 4]
```diff
@@ -126,10 +138,23 @@ func (p *P2PServerV2) SendMessageWithResponse(ctx context.Context, msg *p2pPb.Xu
 	msgOpts := p2p_base.GetMessageOption(opts)
 	filter := p.getFilter(msgOpts)
 	peers, _ := filter.Filter()
+	fmt.Println("=============>peers:", peers)
```
Comment (fxsjy) [inline core/p2p/p2pv2/server.go]: log is better than print

File: core/utxo/utxo_meta.go [hunk 5]
```diff
@@ -166,6 +166,28 @@ func (uv *UtxoVM) GetForbiddenContract() *pb.InvokeRequest {
 	return uv.meta.GetForbiddenContract()
 }
 
+func (uv *UtxoVM) GetGroupChainContract() *pb.InvokeRequest {
```
Comment (yucaowang) [inline core/utxo/utxo_meta.go]: I don't suggest to provide functions for different data in UTXO, it's too heavy. I think we can expose a get/set interface in UTXO and passing the key/value of the data to make UTXO more clear.

File: core/utxo/group_chain.go [hunk 6]
```diff
@@ -0,0 +1,105 @@
...
+	}
+	res := strings.Split(string(target), "\x01")
+	for _, v := range res {
+		ipMap[v] = true
+	}
+
+	return ipMap
+}
+
+func (uv *UtxoVM) getChainInList() map[string]bool {
```
Comment (yucaowang) [inline core/utxo/group_chain.go]: I think chain node group is a xchaincore feature more than UTXO.

File: core/p2p/base/option.go [hunk 7]
```diff
@@ -7,12 +7,19 @@ type MsgOptions struct {
...
 	// compress
-	Compress bool
+	Compress  bool
+	WhiteList map[string]bool
 }
 
 // MessageOption define single option function
 type MessageOption func(*MsgOptions)
 
+func WithWhiteList(whiteList map[string]bool) MessageOption {
```
Comment (yucaowang) [inline core/p2p/base/option.go]: Why not use WithTargetPeerAddrs or WithTargetPeerIDs?

File: core/p2p/p2pv2/server.go [hunk 8]
```diff
@@ -105,6 +105,18 @@ func (p *P2PServerV2) SendMessage(ctx context.Context, msg *p2pPb.XuperMessage,
 	msgOpts := p2p_base.GetMessageOption(opts)
 	filter := p.getFilter(msgOpts)
 	peers, _ := filter.Filter()
+	// 还要做一层过滤(使用白名单过滤)
+	peersRes := []peer.ID{}
```
Comment (yucaowang) [inline core/p2p/p2pv2/server.go]: move this logic into getFilter to avoid duplicate code

File: core/utxo/utxo_meta.go (hunk 5 above)
Comment (ToWorld) [reply core/utxo/utxo_meta.go]: It is same with ForbiddenContract.

File: core/utxo/group_chain.go (hunk 6 above)
Comment (ToWorld) [reply core/utxo/group_chain.go]: Application state is related to utxo, so i support an api based on utxo.

File: core/p2p/base/option.go (hunk 7 above)
Comment (ToWorld) [reply core/p2p/base/option.go]: Different options, different effects.

File: core/core/group_chain.go (hunk 1 above)
Comment (ToWorld) [reply core/core/group_chain.go]: Corrected with a new name of IsPeerInGroupChain

File: core/core/group_chain.go (hunk 3 above)
Comment (ToWorld) [reply core/core/group_chain.go]: Corrected with a readable name of groupChainCacheUpdateWindow

File: core/p2p/p2pv2/server.go (hunk 4 above)
Comment (ToWorld) [reply core/p2p/p2pv2/server.go]: I have ignored the log print.

**Pull request merged**